### PR TITLE
Decouple viewport scrolling speed from FPS

### LINKED ISSUE #34: To make sure the game logic runs on a fixed pace I want to decouple it from FPS
Currently all game logic is FPS based. This should be detached from FPS and become time(r) based.

It looks like Slick provides `delta`'s between two frames. We know how much time has passed and use that to increase timers/etc whatsoever to make sure entities/world things behave on a different pace.

Some info here: 

- http://stackoverflow.com/questions/8282501/how-do-i-work-with-delta-in-slick2d-lwjgl-or-game-programming-in-general
- http://stackoverflow.com/questions/15906500/setting-the-max-and-min-value-for-delta-in-slick2d
- discussion: https://github.com/Maescool/Catacomb-Snatch/issues/979

### PR DESCRIPTION
**Pull request by arjenvanderende**
Fixes #34 

To allow for subpixel movement I've switched from `Vector2D` to `Vector2f` for viewport movement. Structure animation isn't being handled yet.

I dislike `Vector2D` because it _sometimes_ loses precision - due to conversion from float to int - but the interface doesn't make clear when that happens. I think it should be converted to use integers only. If subpixel precision is necessary `Vector2f` should be used instead.

Speaking of `Vector2f`, I dislike that one as well because it's mutable :wink: When adding or scaling the vector it mutates the state of the current instance. This is counter-intuitive for me. For now I've fixed this by explicitly calling `.copy()` where necessary. I'd rather stear clear of mutating state and implement a non-mutable `Vector2f`. What do you think @stefanhendriks?

### CODE REVIEW

**Comment by stefanhendriks** [comment]
I'll look into it into more detail later - but this solves #34 how? Or is this a pre-requisite?

**Comment by arjenvanderende** [comment]
It solves #34 by using the delta provided by Slick.

**Comment by stefanhendriks** [comment]
my bad! I now understand why ... :)

File: src/main/java/com/fundynamic/d2tm/game/map/renderer/MapCellViewportRenderer.java [hunk 1]
```diff
@@ -26,17 +27,17 @@ public MapCellViewportRenderer(Map map, int tileHeight, int tileWidth, Vector2D
         cellsThatFitVertically = (windowDimensions.getY() / tileHeight) + 1;
     }
 
-    public void render(Image imageToDrawOn, Vector2D viewingVector, Renderer<MapCell> renderer) throws SlickException {
-        int startCellX = (viewingVector.getX() / tileWidth);
-        int startCellY = (viewingVector.getY() / tileHeight);
+    public void render(Image imageToDrawOn, Vector2f viewingVector, Renderer<MapCell> renderer) throws SlickException {
+        int startCellX = (int)(viewingVector.getX() / tileWidth);
```
Comment (stefanhendriks) [inline src/main/java/com/fundynamic/d2tm/game/map/renderer/MapCellViewportRenderer.java]: Vector2D wraps `Vector2f` so you could actually change this back to `Vector2D` and use the `getX` and `getY` methods? 

Although then you also would need some way to get the 'original' values (with float). So its a PITA either way.

File: src/main/java/com/fundynamic/d2tm/game/state/PlayingState.java [hunk 2]
```diff
@@ -86,8 +90,9 @@ public void render(GameContainer container, StateBasedGame game, Graphics g) thr
 
     @Override
     public void update(GameContainer container, StateBasedGame game, int delta) throws SlickException {
+        float deltaF = delta / 1000f;
```
Comment (stefanhendriks) [inline src/main/java/com/fundynamic/d2tm/game/state/PlayingState.java]: Although I can make an educated guess, perhaps you can make this explicit that delta is (i think) in miliseconds? And what `deltaF` means?

**Comment by stefanhendriks** [comment]
ok so I found a bit of time ;-) thx for the PR! really nice to get this in quick! :)

File: src/main/java/com/fundynamic/d2tm/game/state/PlayingState.java (hunk 2 above)
Comment (arjenvanderende) [inline src/main/java/com/fundynamic/d2tm/game/state/PlayingState.java]: It's in milliseconds, yes. Renamed to `deltaMs` now.

File: src/main/java/com/fundynamic/d2tm/game/map/renderer/MapCellViewportRenderer.java (hunk 1 above)
Comment (arjenvanderende) [inline src/main/java/com/fundynamic/d2tm/game/map/renderer/MapCellViewportRenderer.java]: If we change `Vector2D` to be more floaty, it becomes even more like `Vector2f`. I don't see why we should implement a custom type when we can use one that is built-in. Besides, some usages of `Vector2D` now depend on in being an integer...

Comment (stefanhendriks) [inline src/main/java/com/fundynamic/d2tm/game/map/renderer/MapCellViewportRenderer.java]: yes, perhaps just get rid of integers all together? get rid of vector2d?

**Comment by stefanhendriks** [comment]
After a bit of thought I think we should keep `Vector2D` and let it wrap this `Vector2f` of Slick. I also don't like this state manipulation and rather have `Vector2D` be an immutable thing that hides the 'internals' of slick vectors. It also is then easier to extend the vector behavior and return casted ints for places where we need to (with `getXAsInt` and `getYAsInt`).

Ie for `scale` we can move that stuff into `Vector2D`. I'll take a stab at this approach now and see how it looks like (so you can check and form an opinion @arjenvanderende). if you don't like it you can always revert the commit ;-)

**Comment by stefanhendriks** [comment]
I have some trouble understanding this piece of code:
```
    @Override
    public void update(GameContainer container, StateBasedGame game, int delta) throws SlickException {
        float deltaMs = delta / 1000f;
        for (Viewport viewport : viewports) {
            viewport.update(deltaMs);
        }
    }
```
The method signature states that `delta` is the time passed in milliseconds between this frame and the last. So `deltaMs` is not the delta, but actually the delta in `seconds` ?

Then when I look at viewport `update` it uses a `scale` method to influence the velocity. Which looks to me as if it uses a `frame per second` reference. Like in the `ViewportMovementListenerTest`  where there is a line calling `update` with `1f` stating that it is `1 frame per second`. 

So we pass the amounts of 'frames' passed in the update method?

**Comment by arjenvanderende** [comment]
@stefanhendriks You're right: the delta is _not_ in milliseconds but in seconds. This way the delta is easy to use, because you can just multiply any vector with it to get the correct fps-independent value.

File: src/main/java/com/fundynamic/d2tm/Game.java [hunk 3]
```diff
@@ -35,7 +37,8 @@ public void initStatesList(GameContainer container) throws SlickException {
                 TILE_HEIGHT
         );
 
-        container.setVSync(true);
+//        container.setVSync(true);
```
Comment (arjenvanderende) [inline src/main/java/com/fundynamic/d2tm/Game.java]: I think we can ditch this line altogether?

**Comment by stefanhendriks** [comment]
I realised the following this morning:

We actually want to say/configure something like this in our code:
`for x time we want to move y (pixels)`

So in the viewport case, we now have:
`for 1 second we move given VELOCITY SPEED`.

I think the division by 1000f should not happen in the PlayingState class but in some sort of `TimedVelocity` class. Something along the lines of:

```
TimedVelocity = timedVelocity(velocity, perMilliseconds); // ie (velocity, 1000f)
Vector2D viewportVector = viewportVector.add(timedVelocity.moveProportionally(deltaInMilliseconds)) // pseudo
```

This means we can re-use the TimedVelocity class for units (where we perhaps want to move things at a different pace, like every 100ms?) and also for animations?

File: src/main/java/com/fundynamic/d2tm/Game.java (hunk 3 above)
Comment (stefanhendriks) [inline src/main/java/com/fundynamic/d2tm/Game.java]: ah yes of course

**Comment by arjenvanderende** [comment]
Why would we want to do that? That way we'll be mixing all different units of speeds, i.e. trikes move at 30 km/h, but infantry moves at 2 m/s. Why would we want to mix use different types of units for this? I feel this makes it more complicated than it needs to be. 

Or do you plan to use this for unit movement?

**Comment by stefanhendriks** [comment]
What i mean is that we should move the conversion from time to movement (per frame) into a class and make it configurable. Because there will be a great diversity in movement speeds per "time unit". Ie flag animation might be per second. But unit movement is per millisecond? Perhaps it is not needed and we can use "per second" as default. 




We could leave it for now and see how flag animation ends up and then see if there is possibly duplication

**Comment by arjenvanderende** [comment]
Structure and selection outline animation is now also correctly timed.

**Comment by stefanhendriks** [comment]
So what is left now to do?

**Comment by arjenvanderende** [comment]
@stefanhendriks I don't think there is much left to do for this PR. Maybe fix the TODO that is introduced? I'd like to keep those to an absolute minimum.

**Comment by stefanhendriks** [comment]
I have fixed the TODO. Ready to merge

**Pull request merged**